### PR TITLE
WIP: testing json patch properties

### PR DESCRIPTION
**Pull request by t0yv0**
# Description

Fixes # (issue)

## Checklist

- [ ] I have added tests that prove my fix is effective or that my feature works

- [ ] I have updated the [CHANGELOG-PENDING](https://github.com/pulumi/pulumi/blob/master/CHANGELOG_PENDING.md) file with my change

- [ ] Yes, there are changes in this PR that warrants bumping the Pulumi Service API version

### CODE REVIEW

**Comment by t0yv0** [comment]
There's some difficulties in json canonicalization. The first piece of bad news is that 

```
jsonpatch is a library which provides functionality for both applying 
[RFC6902 JSON patches](http://tools.ietf.org/html/rfc6902) against documents, 
as well as for calculating & applying [RFC7396 JSON merge patches](https://tools.ietf.org/html/rfc7396).
```

So it can't calculate RFC7396 patches (yet?) - can we please use RFC6902 patches that it can calculate? This would be a change in spec but perhaps okay.

The second set of difficulties is - what constitutes a canonical JSON? I've read up on `json.Marshal()` and while it always produces canonical JSON by sorting map keys, it sorts struct keys by Go field order. This can trip up our use case if client and server do not agree on the versions of struct libs used and there is an accidental reordering, this will trip up checksum calculation.

The quick workaround to that, is memory hungry and quite ridiculous trick of re-encoding the structs into `interface{}` which makes all the maps under the hood and then re-encoding that back. This is now guaranteed canonical.

**Comment by t0yv0** [comment]
The third difficulty is that as random testing shows, the property we want only for documents where maps never contain `null` keys at JSON level. If they do, this trips up exact equality (and hashes would differ). I am hoping that maybe the structure we're serializing will not ever use nulls as map values. If that's not true, we might need to have a canonicalizing pass that drops those nulls.

File: pkg/backend/httpstate/client/json_patch_test.go [hunk 1]
```diff
@@ -0,0 +1,126 @@
...
+}
+
+func decodePatch(raw []byte) (jpatch.Patch, error) {
+	if bytes.Equal(raw, []byte(`{}`)) {
+		return []jpatch.Operation{}, nil
+	}
+	return jpatch.DecodePatch(marshal([]json.RawMessage{raw}))
+}
+
+func TestPatchTurnaround(t *testing.T) {
```
Comment (t0yv0) [inline pkg/backend/httpstate/client/json_patch_test.go:31]: @justinvp you had concerns that the hashes don't match when using JSON patches. This property seems to work for me so far, though models may be a bit unhelpful, tinkering further. IN an abbreviated form, this checks this prop:

```
canonicalizeJson(patch(o, diff(o, m))) == canonicalizeJson(m)
```

**Comment by justinvp** [comment]
One other thought I will throw into the mix. Are we set on using JSON Patch? Chris also experimented with the equivalent of adding newlines to the checkpoint and doing a textual `diff`, which ended up being smaller deltas than JSON Patch.

File: pkg/backend/httpstate/client/json_patch_test.go (hunk 1 above)
Comment (justinvp) [reply pkg/backend/httpstate/client/json_patch_test.go:31]: Thanks for exploring this! It's totally possible that I was doing something silly in the prototype (i.e. not canonicalizing), or I was running into a bug in that library last year.

**Comment by t0yv0** [comment]
Yeah, text diff over canonical json could work if we trust c/s agree on the canonical form.

I feel like we still would like a canonical hash algo that's going to align between client and server so we can exchange these hashes and have high fidelity we're talking about the same object.

Though it's strictly speaking not a requirement. In Chris prototype if we send the full JSON object in the end and don't check any hashes then there's some guarantee things got to the right place on happy path at least.

**Comment by t0yv0** [comment]
Notes to self.

justin/patch branch for prior experiments.

Used github.com/mattbaird/jsonpatch (114*)

Also https://github.com/evanphx/json-patch/issues/56

From comments, there is https://github.com/herkyl/patchwerk (5*)

https://github.com/evanphx/json-patch (785*)

All these give options to support RFC6902 patches but I'd like to similarly test convergence/canonicalization requirements if we need json hashes to match.

**Comment by t0yv0** [comment]
Finding bugs with https://github.com/mattbaird/jsonpatch though maybe that's just how the code uses it. Hmm:

```
        json_patch_test.go:121: [rapid] panic after 0 tests: error in remove for path: '/a/1': Unable to access invalid index: 1: invalid index referenced
            To reproduce, specify -run="TestRFC6902PatchTurnaround/restricted-3" -rapid.failfile="TestRFC6902PatchTurnaround_restricted-3-20220909144300-56171.fail"
            Traceback:
                /Users/anton/pulumi/pkg/backend/httpstate/client/json_patch_test.go:94 in github.com/pulumi/pulumi/pkg/v3/backend/httpstate/client.TestRFC6902PatchTurnaround.func2
                /Users/anton/pulumi/pkg/backend/httpstate/client/json_patch_test.go:147 in github.com/pulumi/pulumi/pkg/v3/backend/httpstate/client.checkTurnaround
                /Users/anton/pulumi/pkg/backend/httpstate/client/json_patch_test.go:131 in github.com/pulumi/pulumi/pkg/v3/backend/httpstate/client.checkTurnaroundThoroughly.func2.1
                /Users/anton/go/pkg/mod/pgregory.net/rapid@v0.4.7/engine.go:276 in pgregory.net/rapid.checkOnce
            Failed test output:
        json_patch_test.go:137: [rapid] draw original: json.RawMessage{0x7b, 0x22, 0x61, 0x22, 0x3a, 0x5b, 0x6e, 0x75, 0x6c, 0x6c, 0x2c, 0x6e, 0x75, 0x6c, 0x6c, 0x5d, 0x7d}
        json_patch_test.go:138: [rapid] draw modified: json.RawMessage{0x7b, 0x22, 0x61, 0x22, 0x3a, 0x5b, 0x5d, 0x7d}
        json_patch_test.go:140: original      = {"a":[null,null]}
        json_patch_test.go:141: modified      = {"a":[]}
        json_patch_test.go:145: patch         = [{"op":"remove","path":"/a/0"},{"op":"remove","path":"/a/1"}]
```

**Comment by lunaris** [comment]
Closing as it looks like this was either superseded by #10788, or an experiment that was part of the (now marked completed) https://github.com/pulumi/home/issues/2308

**Pull request closed without merging**